File: packages/grep/src/suggest.ts
import path from "node:path";
import type { ReadOperations } from "@agent-sh/harness-core";
import { FUZZY_SUGGESTION_LIMIT } from "./constants.js";

/**
 * Return up to FUZZY_SUGGESTION_LIMIT entries in the parent directory that
 * look like the missing basename, sorted most-similar-first. Empty array on
 * any IO error — NOT_FOUND without siblings is still a valid shape.
 *
 * Mirrors the Read tool's implementation; duplicated rather than shared so
 * grep does not depend on the read package.
 */
export async function suggestSiblings(
  ops: ReadOperations,
  missingPath: string,
): Promise<readonly string[]> {
  const dir = path.dirname(missingPath);
  const base = path.basename(missingPath).toLowerCase();
  let entries: readonly string[];
  try {
    entries = await ops.readDirectory(dir);
  } catch {
    return [];
  }
  const scored: { p: string; score: number }[] = [];
  for (const entry of entries) {
    const lower = entry.toLowerCase();
    const score = similarity(base, lower);
    if (score > 0) scored.push({ p: path.join(dir, entry), score });
  }
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, FUZZY_SUGGESTION_LIMIT).map((s) => s.p);
}

function similarity(a: string, b: string): number {
  if (a === b) return 1000;
  if (a.length === 0 || b.length === 0) return 0;
  if (a.includes(b) || b.includes(a)) return 500;
  const prefix = commonPrefix(a, b);
  if (prefix >= 3) return 200 + prefix;
  if (prefix >= 2 && Math.abs(a.length - b.length) <= 2) return 100 + prefix;
  const aExt = extOf(a);
  const bExt = extOf(b);
  if (aExt && aExt === bExt) return 10;
  return 0;
}

function commonPrefix(a: string, b: string): number {
  const n = Math.min(a.length, b.length);
  let i = 0;
  while (i < n && a[i] === b[i]) i++;
  return i;
}

function extOf(name: string): string {
  const dot = name.lastIndexOf(".");
  return dot > 0 ? name.slice(dot) : "";
}

File: crates/grep/src/suggest.rs
use crate::constants::FUZZY_SUGGESTION_LIMIT;
use std::path::Path;

/// Return up to [`FUZZY_SUGGESTION_LIMIT`] entries in the parent
/// directory of `missing_path` that look like its basename, sorted
/// most-similar first. Same algorithm as the TS `suggestSiblings`.
pub fn suggest_siblings(missing_path: &str) -> Vec<String> {
    let p = Path::new(missing_path);
    let parent = match p.parent() {
        Some(pp) => pp,
        None => return Vec::new(),
    };
    let base = p
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_lowercase();
    let entries = match std::fs::read_dir(parent) {
        Ok(e) => e,
        Err(_) => return Vec::new(),
    };
    let mut scored: Vec<(i64, String)> = entries
        .filter_map(|e| e.ok())
        .filter_map(|e| {
            e.file_name()
                .to_str()
                .map(|n| n.to_string())
        })
        .filter_map(|name| {
            let s = similarity(&base, &name.to_lowercase());
            if s > 0 {
                Some((s, parent.join(&name).to_string_lossy().into_owned()))
            } else {
                None
            }
        })
        .collect();
    scored.sort_by(|a, b| b.0.cmp(&a.0));
    scored
        .into_iter()
        .take(FUZZY_SUGGESTION_LIMIT)
        .map(|(_, p)| p)
        .collect()
}

fn similarity(a: &str, b: &str) -> i64 {
    if a == b {
        return 1000;
    }
    if a.is_empty() || b.is_empty() {
        return 0;
    }
    if a.contains(b) || b.contains(a) {
        return 500;
    }
    let prefix = common_prefix(a, b);
    if prefix >= 3 {
        return 200 + prefix as i64;
    }
    let a_len = a.len() as i64;
    let b_len = b.len() as i64;
    if prefix >= 2 && (a_len - b_len).abs() <= 2 {
        return 100 + prefix as i64;
    }
    let a_ext = ext_of(a);
    let b_ext = ext_of(b);
    if !a_ext.is_empty() && a_ext == b_ext {
        return 10;
    }
    0
}

fn common_prefix(a: &str, b: &str) -> usize {
    a.bytes().zip(b.bytes()).take_while(|(x, y)| x == y).count()
}

fn ext_of(name: &str) -> &str {
    match name.rfind('.') {
        Some(i) if i > 0 => &name[i..],
        _ => "",
    }
}
